### PR TITLE
Add RepaymentRouter

### PR DESCRIPTION
**Pull request by jdkanani**

### CODE REVIEW

File: contracts/RepaymentRouter.sol [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  {
+    require(token != address(0));
+    require(amount > 0);
+
+    // Get registry entry and check if entry is valid
+    var (version, creditor, termsContract, termsContractParameters) = debtRegistry.get(entryHash);
+    require(creditor != address(0));
+
+    // Transfer amount to creditor
+    DebtToken debtToken = DebtToken(token);
```
Comment (nadavhollander) [inline contracts/RepaymentRouter.sol]: This is incorrect, but I can see how you got confused.  My fault for not explaining this clearly enough, sorry -- the DebtToken file is unrelated to the RepaymentRouter.  The token address in the `repay` function can be any token (either ERC20 / [ERC721](https://github.com/ethereum/EIPs/issues/721)).

FYI DebtToken is related to other functionality in Dharma protocol -- i.e. managing business logic for the actual tokens that represent ownership in any given debt.

If ERC721, would probably need to be a separate function (i.e. `repayNFT`)

File: contracts/base/Ownable.sol [hunk 2]
```diff
@@ -10,6 +10,9 @@ pragma solidity 0.4.18;
 contract Ownable {
     address public owner;
 
+    // transfer ownership event
+    event OwnershipTransferred(address indexed previousOwner, address indexed newOwner);
```
Comment (nadavhollander) [inline contracts/base/Ownable.sol]: 👍

File: contracts/RepaymentRouter.sol [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    debtToken.transferFrom(msg.sender, creditor, amount);
+
+    // Notify terms contract
+    // TermsContract termsContract = TermsContract(termsContract);
+    // termsContract.registerRepayment(msg.sender, termsContractParameters, amount, token);
```
Comment (nadavhollander) [inline contracts/RepaymentRouter.sol]: This looks great!  Need to add the actual TermsContractInterface to the repo (https://whitepaper.dharma.io/#terms-contract-interface), but this should be good.

**Comment by jdkanani** [comment]
I'll fix those issues soon.

File: contracts/RepaymentRouter.sol (hunk 1 above)
Comment (jdkanani) [reply contracts/RepaymentRouter.sol]: You are right. `repayNFT` makes much more sense.

Comment (jdkanani) [reply contracts/RepaymentRouter.sol]: Plus, I will add ERC20Token file for `repay`

**Comment by jdkanani** [comment]
I haven't changed `DebtToken`, but added new `ERC20Token` (easy to get idea from file name).  Let me know if that's okay?

Regarding `repayNFT` and `termsContract`, I've asked few questions on rocket chat.

File: contracts/RepaymentRouter.sol [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+*/
+
+pragma solidity 0.4.18;
+
+import "./base/ERC20Token.sol";
+import "./base/NonFungibleToken.sol";
+import "./DebtRegistry.sol";
+import "./TermsContract.sol";
+
+contract RepaymentRouter {
```
Comment (nadavhollander) [inline contracts/RepaymentRouter.sol]: Nit: we use 4-space indents, not 2-space indents

File: contracts/RepaymentRouter.sol [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    var (, creditor, termsContract,) = debtRegistry.get(entryHash);
+    require(creditor != address(0));
+
+    // check if terms contract param is same as registry record
+    bytes32 paramsHash = debtRegistry.getTermsContractParametersHash(entryHash);
+    require(paramsHash == keccak256(termsContractParameters));
+
+    // Transfer NFT to creditor
+    NonFungibleToken tokenInstance = NonFungibleToken(token);
+    tokenInstance.transferFrom(msg.sender, creditor, tokenId);
```
Comment (nadavhollander) [inline contracts/RepaymentRouter.sol]: Should check `implementsERC721` flag as a precaution -- the `transferFrom` call signature is nearly identical in ERC20 and ERC721.  Could create weird edge cases and attacks on the integrity of the repayments registered with the terms contract.

**Comment by nadavhollander** [comment]
Looks awesome!  _Almost_ all of the way there -- added one comment for `repayNFT` that should be a quick fix.

Also -- there's a bunch of linting issues -- if you run `yarn lint` or `npm run lint` you should see them.

Thanks a bunch!

File: contracts/RepaymentRouter.sol (hunk 5 above)
Comment (jdkanani) [reply contracts/RepaymentRouter.sol]: Good point. Fixed and pushed.

**Comment by jdkanani** [comment]
Fixed and pushed lint error and added `repayNFT` flag check.

**Comment by jdkanani** [comment]
Thanks @nadavhollander

**Pull request merged**